oxenss/snode/service_node.cpp

```diff
@@ -43,29 +44,426 @@ using MISSING_PUBKEY_THRESHOLD = std::ratio<3, 100>;
 /// TODO: there should be config.h to store constants like these
 constexpr auto OXEND_PING_INTERVAL = 30s;
 
+// How often to trigger 'check_new_members' which checks for 'data ready' handshakes from
+// swarm members and propagate a DB dump if necessary.
 constexpr auto NEW_SWARM_MEMBER_INTERVAL = 10s;
 
+struct SerialiseRetryableRequestsResult {
+    SerialiseBTResult bt;
+    std::vector<RequestRetry> retryable_requests;
+};
+
+SerialiseDataReadyRequestResult serialise_data_ready_request(
+        Serialise serialise, std::string_view read_data, const DataReadyRequest& write_data) {
+    SerialiseDataReadyRequestResult result = {};
+    uint32_t version = 0;
+    constexpr std::string_view VERSION_KEY = "@";
+    constexpr std::string_view STATUS_KEY = "s";
+    constexpr std::string_view NEED_DB_DUMP_KEY = "t";
+    static_assert(VERSION_KEY < STATUS_KEY);
+    static_assert(STATUS_KEY < NEED_DB_DUMP_KEY);
+
+    if (serialise == Serialise::Write) {
+        oxenc::bt_dict_producer d;
+        d.append(VERSION_KEY, version);
+        d.append(NEED_DB_DUMP_KEY, write_data.needs_db_dump);
+        result.bt.write_payload = d.view();
```

Comment:
Avoid copying the string with:
```suggestion
        result.bt.write_payload = std::move(d).str();
```